cmd/collect-profiles/main.go

```diff
@@ -0,0 +1,364 @@
+		},
+	}
+}
+
+func verifyCertAndKeyExist(certPath, keyPath string) error {
+	fi, err := os.Stat(certPath)
+	if err != nil {
+		return err
+	}
+	if fi.Size() == 0 {
+		return fmt.Errorf("cert file should not be empty")
+	}
+
+	fi, err = os.Stat(keyPath)
+	if err != nil {
+		return err
+	}
+	if fi.Size() == 0 {
+		return fmt.Errorf("key file should not be empty")
+	}
+	return nil
+}
+
+func separateConfigMapsIntoNewestAndExpired(configMaps []corev1.ConfigMap) (newestCMs []corev1.ConfigMap, expiredCMs []corev1.ConfigMap) {
+	// Group ConfigMaps by GenerateName
+	newestConfigMaps := map[string]corev1.ConfigMap{}
+	for _, cm := range configMaps {
+		if _, ok := newestConfigMaps[cm.GenerateName]; !ok {
+			newestConfigMaps[cm.GenerateName] = cm
+			continue
+		}
+		if cm.CreationTimestamp.After(newestConfigMaps[cm.GenerateName].CreationTimestamp.Time) {
+			newestConfigMaps[cm.GenerateName], cm = cm, newestConfigMaps[cm.GenerateName]
+		}
+		expiredCMs = append(expiredCMs, cm)
+	}
+
+	for _, v := range newestConfigMaps {
+		newestCMs = append(newestCMs, v)
+	}
+
+	return newestCMs, expiredCMs
+}
+
+type argument struct {
+	generateName string
+	url          *url.URL
+}
+
+func newArgument(s string) (*argument, error) {
+	splitStrings := strings.SplitN(s, ":", 2)
+	if len(splitStrings) != 2 {
+		return nil, fmt.Errorf("%s is an invalid argument, should match configMapName:url", s)
+	}
+
+	url, err := url.Parse(splitStrings[1])
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.ToLower(url.Scheme) != "https" {
+		return nil, fmt.Errorf("URL Scheme must be HTTPS")
+	}
+
+	arg := &argument{
+		generateName: splitStrings[0],
+		url:          url,
+	}
+
+	return arg, nil
+}
+
+func getHttpClient(certPath, keyPath string) (*http.Client, error) {
+	cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+				Certificates:       []tls.Certificate{cert},
+			},
+		},
+	}, nil
+}
+
+func requestURLBody(httpClient *http.Client, u *url.URL) ([]byte, error) {
+	response, err := httpClient.Do(&http.Request{
+		Method: http.MethodGet,
+		URL:    u,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s responded with %d status code instead of %d", u, response.StatusCode, http.StatusOK)
+	}
+
+	var b bytes.Buffer
+	if _, err := io.Copy(&b, response.Body); err != nil {
+		return nil, fmt.Errorf("error reading response body: %v", err)
+	}
+
+	return b.Bytes(), nil
+}
+
+func populateServingCert(ctx context.Context, client client.Client) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pprof-serving-secret",
+			Namespace: "openshift-operator-lifecycle-manager",
+		},
+	}
+	err := client.Get(ctx, types.NamespacedName{Namespace: "openshift-operator-lifecycle-manager", Name: "pprof-cert"}, secret)
```

Comment:
Would it make sense for either of these namespace or secret name values to be configurable that default to the OLM and "pprof-secret" values? If not, these values could be live as constant variables in this package.

Reply:
Let's go with constants given that no other project will probably ever use this binary.